File: client-sdk/go/types/transaction.go
// TODO: Move this package to the Go client-sdk.
package types

import (
	"fmt"

	"github.com/oasisprotocol/oasis-core/go/common/cbor"

	"github.com/oasisprotocol/oasis-bridge/client-sdk/go/crypto/signature"
)

// Transaction is a runtime transaction.
type Transaction struct {
	cbor.Versioned

	Call       Call     `json:"call"`
	AuthInfo   AuthInfo `json:"ai"`
	Signatures [][]byte `json:"sigs"`
}

// ValidateBasic performs basic validation on the transaction.
func (t *Transaction) ValidateBasic() error {
	if t.V != 1 {
		return fmt.Errorf("transaction: unsupported version")
	}
	if len(t.Signatures) == 0 {
		return fmt.Errorf("transaction: malformed transaction")
	}
	if len(t.Signatures) != len(t.AuthInfo.SignerInfo) {
		return fmt.Errorf("transaction: malformed transaction")
	}
	return nil
}

// Call is a method call.
type Call struct {
	Method string          `json:"method"`
	Body   cbor.RawMessage `json:"body"`
}

// AuthInfo contains transaction authentication information.
type AuthInfo struct {
	SignerInfo []SignerInfo `json:"si"`
	Fee        Fee          `json:"fee"`
}

// Fee contains the transaction fee information.
type Fee struct {
	Amount BaseUnits `json:"amount"`
	Gas    uint64    `json:"gas"`
}

// SignerInfo contains transaction signer information.
type SignerInfo struct {
	PublicKey signature.PublicKey `json:"pub"`
	Nonce     uint64              `json:"nonce"`
}

// CallResult is the method call result.
type CallResult struct {
	Ok     cbor.RawMessage   `json:"ok,omitempty"`
	Failed *FailedCallResult `json:"fail,omitempty"`
}

// IsSuccess checks whether the call result indicates success.
func (cr *CallResult) IsSuccess() bool {
	return cr.Failed == nil
}

// FailedCallResult is a failed call result.
type FailedCallResult struct {
	Module string `json:"module"`
	Code   uint32 `json:"code"`
}

// String returns the string representation of a failed call result.
func (cr FailedCallResult) String() string {
	return fmt.Sprintf("module: %s code: %d", cr.Module, cr.Code)
}

File: runtime-sdk/src/types/transaction.rs
//! Transaction types.
use serde::{Deserialize, Serialize};
use thiserror::Error;

use oasis_core_runtime::common::cbor;

use crate::{
    crypto::signature::{PublicKey, Signature},
    types::token,
};

/// Error.
#[derive(Error, Debug)]
pub enum Error {
    #[error("unsupported version")]
    UnsupportedVersion,
    #[error("malformed transaction")]
    MalformedTransaction,
}

/// Transaction.
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct Transaction {
    #[serde(rename = "v")]
    pub version: u16,

    #[serde(rename = "call")]
    pub call: Call,

    #[serde(rename = "ai")]
    pub auth_info: AuthInfo,

    #[serde(rename = "sigs")]
    pub signatures: Vec<Signature>,
}

impl Transaction {
    /// Perform basic validation on the transaction.
    pub fn validate_basic(&self) -> Result<(), Error> {
        if self.version != 1 {
            return Err(Error::UnsupportedVersion);
        }
        if self.signatures.is_empty() {
            return Err(Error::MalformedTransaction);
        }
        if self.auth_info.signer_info.len() != self.signatures.len() {
            return Err(Error::MalformedTransaction);
        }
        Ok(())
    }
}

/// Method call.
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct Call {
    #[serde(rename = "method")]
    pub method: String,

    #[serde(rename = "body")]
    pub body: cbor::Value,
}

/// Transaction authentication information.
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct AuthInfo {
    #[serde(rename = "si")]
    pub signer_info: Vec<SignerInfo>,

    #[serde(rename = "fee")]
    pub fee: Fee,
}

/// Transaction fee.
#[derive(Clone, Debug, Default, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct Fee {
    #[serde(rename = "amount")]
    pub amount: token::BaseUnits,

    #[serde(rename = "gas")]
    pub gas: u64,
}

/// Transaction signer information.
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct SignerInfo {
    #[serde(rename = "pub")]
    pub public_key: PublicKey,

    #[serde(rename = "nonce")]
    pub nonce: u64,
}

/// Call result.
#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub enum CallResult {
    #[serde(rename = "ok")]
    Ok(cbor::Value),

    #[serde(rename = "fail")]
    Failed {
        #[serde(rename = "module")]
        module: String,

        #[serde(rename = "code")]
        code: u32,
    },
}
